ecommerce/models.py

```diff
@@ -154,3 +156,97 @@ def save(self, *args, **kwargs):
     def __str__(self):
         """Description for CoursePrice"""
         return "CoursePrice for {}, price={}, is_valid={}".format(self.course_run, self.price, self.is_valid)
+
+
+class Coupon(Model):
+    """
+    Model for a coupon. This stores the discount for the coupon, how many people can use it, and how many times
+    a person can use it.
+
+    When a coupon is redeemed by a purchaser the counter on this object is decremented
+    and a UserCoupon object is created for that particular purchaser.
+    """
+    PERCENT_DISCOUNT = 'percent-discount'
+    FIXED_DISCOUNT = 'fixed-discount'
+
+    AMOUNT_TYPES = [PERCENT_DISCOUNT, FIXED_DISCOUNT]
+
+    COURSE_RUN = 'course-run'
+    COURSE = 'course'
+    PROGRAM = 'program'
+    PRODUCT_TYPES = [PROGRAM, COURSE, COURSE_RUN]
+
+    # The coupon code used for redemption by the purchaser in the user interface.
+    # If blank, the purchaser may not redeem this coupon through the user interface,
+    # though it may be redeemed in their name by an administrator.
+    coupon_code = TextField(null=True, blank=True)
+    product_type = CharField(
+        choices=[(_type, _type) for _type in PRODUCT_TYPES],
+        max_length=30,
+    )
+    product_id = IntegerField(null=False)
+    # percent or fixed discount
+    amount_type = CharField(
+        choices=[(_type, _type) for _type in AMOUNT_TYPES],
+        max_length=30,
+    )
+    # Either a number from 0 to 1 representing a percent, or a fixed value for discount
+    amount = DecimalField(decimal_places=2, max_digits=20)
+
+    # Number of people this coupon can be redeemed by
+    num_coupons_available = IntegerField(null=False)
+    # Number of times a person can redeem a coupon
+    num_redemptions = IntegerField(null=False)
+    # After this time the coupons will not be redeemable
+    expiration_date = DateTimeField(null=True)
+    # If true, coupons are not presently redeemable
+    disabled = BooleanField(default=False)
+
+    def __str__(self):
+        """Description for Coupon"""
+        return "Coupon {amount_type} {amount} for {product_type} {product_id}, {num_coupons_available} left".format(
+            amount_type=self.amount_type,
+            amount=self.amount,
+            product_type=self.product_type,
+            product_id=self.product_id,
+            num_coupons_available=self.num_coupons_available,
+        )
+
+
+class UserCoupon(Model):
+    """
+    Model for uses of a coupon by a user.
+    """
+    coupon = ForeignKey(Coupon, on_delete=SET_NULL)
+    user = ForeignKey(settings.AUTH_USER_MODEL, on_delete=SET_NULL)
+
+    # Number of times left the user can use the coupon in a purchase
+    available_redemptions = IntegerField()
```

Comment:
We can make the field read-only in the admin but there's not much we can do to prevent people from modifying it in the shell if they want.

Speaking of, I'll add admin interfaces for these models